pkg/mei/mei_linux.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2020 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package mei implements a wrapper on top of Linux's MEI (Intel ME Interface,
+// formerly known as HECI). This module requires Linux, and the `mei_me` driver.
+// Once loaded, this driver will expose a `/dev/mei0` device, that can be
+// accessed through this library.
+package mei
+
+import (
+	"encoding/binary"
+	"fmt"
+	"os"
+	"syscall"
+	"unsafe"
+
+	"github.com/vtolstov/go-ioctl"
+)
+
+// DefaultMEIDevicePath is the path of the default MEI device. This file will be
+// present if the "mei_me" kernel module is loaded.
+var DefaultMEIDevicePath = "/dev/mei0"
+
+// HECIGuids maps the known HECI GUIDs to their values. The MEI interface wants
+// little-endian. See all the GUIDs at
+// https://github.com/intel/lms/blob/master/MEIClient/Include/HECI_if.h
+var (
+	// "8e6a6715-9abc-4043-88ef-9e39c6f63e0f"
+	MKHIGuid = ClientGUID{0x15, 0x67, 0x6a, 0x8e, 0xbc, 0x9a, 0x43, 0x40, 0x88, 0xef, 0x9e, 0x39, 0xc6, 0xf6, 0x3e, 0xf}
+)
+
+// see include/uapi/linux/mei.h
+var (
+	IoctlMEIConnectClient = ioctl.IOWR('H', 0x01, uintptr(len(ClientGUID{})))
+)
+
+// ClientGUID is the data buffer to pass to `ioctl` to connect to
+// MEI. See include/uapi/linux/mei.h .
+type ClientGUID [16]byte
+
+// ClientProperties is the data buffer returned by `ioctl` after connecting to
+// MEI. See include/uapi/linux/mei.h .
+type ClientProperties [6]byte
+
+// MaxMsgLength is the maximum size of a message for this client.
+func (c ClientProperties) MaxMsgLength() uint32 {
+	return binary.LittleEndian.Uint32(c[:4])
+}
+
+// ProtocolVersion is this client's protocol version.
+func (c ClientProperties) ProtocolVersion() uint8 {
+	return c[4]
+}
+
+// MEI represents an Intel ME Interface object.
+type MEI struct {
+	fd               *os.File
+	ClientProperties ClientProperties
+}
+
+// Open opens the specified MEI device, using the client type defined by the
+// given name. See `HECIGuids` in this package.
+func (m *MEI) Open(meiPath string, guid ClientGUID) error {
+	fd, err := os.OpenFile(meiPath, os.O_RDWR, 0755)
```

Comment:
os.OpenFile will still set the FD to be non-blocking and add it to the epoll, despite using the sys Write and Read below. You want it to be a blocking fd, I assume.

To do that, do a syscall.Open and use os.NewFile to make a *os.File out of it. Then you should be even able to use File.Read/Write again.

Reply:
good point - we tested the code and it actually does what we expect, but uniforming on the same API is better

Reply:
done